### ISSUE TITLE
Null Value error for 'path'

### ISSUE THREAD

**Issue by damoose** [NathanQuellec/WSL-Studio#115]
Installed the lates WSL Studio from the MS Store.

Tried to add the Linux Mint Debian Edition

Distribution Name: lmde
Creation Mode: Docker Hub
Docker Hub Repository: linuxmintd/lmde6-amd64:latest

Create results in an error:
WSL Studio has failed to create your distribution
Value cannot be null. (Parameter 'path')

It is unclear to me what I have missed in the documentation.

**Comment by NathanQuellec** [comment]
Hello,
Thank you for raising this issue !

On my side, I'm able to create a distribution from your image
Can you please tell me if the distribution is displayed when you enter "wsl -l" command on your Windows terminal ?

Can you also provide me the application log located in 

`C:\Users\YOUR_USER_PROFILE_NAME\AppData\Local\Packages\YOUR_WSL_STUDIO_PACKAGE_NAME\LocalState\WslStudio\.log`

My guess is that WSL Studio cannot resolve one of the path used to store the distribution files (in `
C:\Users\YOUR_USER_PROFILE_NAME\AppData\Roaming\WslStudio`). 

Thank you for your help to improve the project 👍

**Comment by damoose** [comment]
Hello,

Thank you for responding.

wsl - l reports:

Ubuntu (Default)
docker-desktop

If I understand correctly you want the log files from the WSL application that I installed from the MS Store.

C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState

This directory is empty. There are no logs.

This directory C:\Users\YOUR_USER_PROFILE_NAME\AppData\Roaming\WslStudio does not exist.

**Comment by NathanQuellec** [comment]
Ok, for some reason, the app cannot create folders or files on your system. Can you provide some information about it ? (like the Windows version)

Try one of those workarounds : 

- Start WSL Studio with admin rights
or
- Create manually the "WslStudio" directory inside the "Roaming" directory. In addition, create three directories inside LocalState directory : "WslStudio" , "WslStudio\.log" and "WslStudio\.tmp" 

----------------------------------------

Development to do (for me or others contributors) :  

- Enhance error message by warning the user that app directories are not created
- Investigate why directories cannot be created on some systems

**Comment by damoose** [comment]
Tried both workarounds. Both failed with the original error.

I created these directories:
C:\Users\\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio\LocalState\WslStudio
C:\Users\\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio\LocalState\WslStudio.log
C:\Users\\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio\LocalState\WslStudio.tmp

The log and tmp folders have no data.

I rebooted the PC and ran the tests again as specified.

No joy. Still the same null value error.

**Comment by NathanQuellec** [comment]
Sorry there was a formatting issue in my previous answer.
Below the list of **directories** to create :  
`C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\WslStudio`

`C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\WslStudio\.log`

`C:\UsC:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\WslStudio\.tmp`

`C:\Users\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio`

**Comment by damoose** [comment]
So still not successful.
I ran both as user and administrator.

Here are the directorates:
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\AC
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\AppData
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalCache
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\RoamingState
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\Settings
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\SystemAppData
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\TempState
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\AC\Temp
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalCache\Local
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalCache\Roaming
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalCache\Local\Microsoft
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalCache\Roaming\Microsoft
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalCache\Roaming\Microsoft\Windows
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalCache\Roaming\Microsoft\Windows\Start Menu
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalCache\Roaming\Microsoft\Windows\Start Menu\Programs

C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\.log
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\.tmp
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\WslStudio

C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\Settings\roaming.lock
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\Settings\settings.dat
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\SystemAppData\Helium
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\SystemAppData\Helium\User.dat
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\SystemAppData\Helium\UserClasses.dat

C:\Users\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio

The .log dir is empty. Is there another way to get more detail during the run?

There is another message that appears before the error, but goes by too fast.  And when I close the error dialog, there is nothing there. I'll try OBS and see if I can get a cap of it.

**Comment by NathanQuellec** [comment]
Why those two lines ? : 
```
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\.log
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\.tmp
```

It's this : 
```
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\WslStudio\.log
C:\Users\MY_USER_PROFILE_NAME\AppData\Local\Packages\14543NathanQuellec.WSLStudio_yzeq27d2nx5vm\LocalState\WslStudio\.tmp
```

Do you use Windows 10 or 11 ? I will try to debug on a system the most close of yours

Unfortunaly, logs are the only way to get details during the runtime.
Except if you are willing to debug the project by putting a breakpoint (in App.xaml.cs for this issue)

**Comment by damoose** [comment]
Apologies. I misread the original.

I moved the directories under WslStudio. The error persists and no logs are created.

According to winver I am running:
Windows 11
23H2 (OS Build 22631.4112)

Fresh clean install about 2 weeks ago... so not much cruft.

Been a while since I ran VS and I don't have a personal license but if we can figure it out... What the heck.

**Comment by damoose** [comment]
So got VS installed.
Cloned your repo.
Changed the output directory to my local dirs.

Had to change the configuration in the Configuration Manager to x64 from AnyCPU to get the solution to build.

**Comment by damoose** [comment]
I put break points in App.xaml.cs at lines:

70 - never reached
200
217
283

In program.cs line 20

I stepped through line by line and the following exceptions occurred:

System.Exception
  HResult=0x80131500
  Message=Failed to fetch image layers
  Source=WSL Studio
  StackTrace:
   at WSLStudio.Helpers.DockerHelper.<GetLayers>d__14.MoveNext() in D:\src\WSL-Studio\WSLStudio\Helpers\DockerHelper.cs:line 332
   at WSLStudio.Services.Factories.DockerHubDistributionFactory.<CreateDistribution>d__0.MoveNext() in D:\src\WSL-Studio\WSLStudio\Services\Factories\DockerHubDistributionFactory.cs:line 43

  This exception was originally thrown at this call stack:
    WSLStudio.Helpers.DockerHelper.GetLayers(WSLStudio.Models.Docker.AuthToken, WSLStudio.Contracts.Models.Docker.Manifests.IImageManifest, string) in DockerHelper.cs
    WSLStudio.Services.Factories.DockerHubDistributionFactory.CreateDistribution(string, string, string) in DockerHubDistributionFactory.cs
    
    System.ArgumentNullException
  HResult=0x80004003
  Message=Value cannot be null. (Parameter 'path')
  Source=System.Private.CoreLib
  StackTrace:
   at System.ArgumentNullException.Throw(String paramName)
   at System.IO.DirectoryInfo..ctor(String path)
   at WSLStudio.Helpers.FilesHelper.RemoveDirContent(String dirPath) in D:\src\WSL-Studio\WSLStudio\Helpers\FilesHelper.cs:line 48
   at WSLStudio.Services.Factories.DockerHubDistributionFactory.<CreateDistribution>d__0.MoveNext() in D:\src\WSL-Studio\WSLStudio\Services\Factories\DockerHubDistributionFactory.cs:line 67
   at WSLStudio.Services.DistributionService.<CreateDistribution>d__8.MoveNext() in D:\src\WSL-Studio\WSLStudio\Services\DistributionService.cs:line 112

  This exception was originally thrown at this call stack:
    [External Code]
    WSLStudio.Helpers.FilesHelper.RemoveDirContent(string) in FilesHelper.cs
    WSLStudio.Services.Factories.DockerHubDistributionFactory.CreateDistribution(string, string, string) in DockerHubDistributionFactory.cs
    WSLStudio.Services.DistributionService.CreateDistribution(string, string, string) in DistributionService.cs

System.ArgumentNullException
  HResult=0x80004003
  Message=Value cannot be null. (Parameter 'path')
  Source=System.Private.CoreLib
  StackTrace:
   at System.ArgumentNullException.Throw(String paramName)
   at System.IO.DirectoryInfo..ctor(String path)
   at WSLStudio.Helpers.FilesHelper.RemoveDirContent(String dirPath) in D:\src\WSL-Studio\WSLStudio\Helpers\FilesHelper.cs:line 48
   at WSLStudio.Services.Factories.DockerHubDistributionFactory.<CreateDistribution>d__0.MoveNext() in D:\src\WSL-Studio\WSLStudio\Services\Factories\DockerHubDistributionFactory.cs:line 67
   at WSLStudio.Services.DistributionService.<CreateDistribution>d__8.MoveNext() in D:\src\WSL-Studio\WSLStudio\Services\DistributionService.cs:line 112
   at WSLStudio.ViewModels.DistrosListDetailsVM.<CreateDistributionViewModel>d__64.MoveNext() in D:\src\WSL-Studio\WSLStudio\ViewModels\DistrosListDetailsVM.cs:line 439

  This exception was originally thrown at this call stack:
    [External Code]
    WSLStudio.Helpers.FilesHelper.RemoveDirContent(string) in FilesHelper.cs
    WSLStudio.Services.Factories.DockerHubDistributionFactory.CreateDistribution(string, string, string) in DockerHubDistributionFactory.cs
    WSLStudio.Services.DistributionService.CreateDistribution(string, string, string) in DistributionService.cs
    WSLStudio.ViewModels.DistrosListDetailsVM.CreateDistributionViewModel(string, string, string) in DistrosListDetailsVM.cs

**Comment by NathanQuellec** [comment]
First of all, thank you for your dedication ! It is really helping the project

I'm developing on Windows so may be I missed something about Windows 11 compatibility

To give you some insights of the code : 

1. At startup, "OnLaunched"  method in line 261 of App.xaml.cs is running various check to detect if WSL is enabled on the system **and** it is creating the app folders. (you should debug this method and not program.cs)
![image](https://github.com/user-attachments/assets/57831e83-4889-4171-831e-0da853ca81d3)

3. Once the apps folders are created, you can create a custom distribution. When you download an image from Docker Hub, the different layers of this image will be stored as .tar files in the .tmp folder. At the end, all files will be merged to create the resulting filesystem of the distribution.

If you never reach the CreateProjectFolders method (line 70 as you mentionned), the properties AppDirPath, TmpDirPath and LogDirPath are never set. Those are used in every part of this project for reasons mentioned just before. It's why you have the exception on RemoveDirContent because it's trying to clean the temp directory using the path set in TmpDirPath object

The issue is maybe from a bad asynchronous management of the method CreateProjectFolders . 
Or another error before the call of the method

**Comment by damoose** [comment]
OK. So I removed the Program.cs break point.

Changed the break points in App.xaml.cs, lines:

263
277
283

Added a break point in WslHelper line 45.

F5 then stepping through:

App.xaml.cs
263 -> completed
264 -> completed with a black app rectangle created.
266 -> Off to WslHelper line 45

In WslHelper.cs
45 -> completes
53 -> completes and I have the full WSL app displayed
54 -> waits until I input all the information, then fails

It never returns to App.xmal.cs and hence never creates the directories.

Seems to me that you should create the directories and the logger before you attempt anything else.

**Comment by NathanQuellec** [comment]
You are right about creating the directories and logger in first.
I create a new branch **feature/Issue-115_Fix_Dir_creation** with the fix. Can you tell me if it's working now when you run the project ?

It also seems that the method CheckWslMicrosoftStore in WslHelper is blocking on your system.
Can you run the command **winget ls  -q 'Windows Subsystem for Linux'** on powershell and tell me what is displayed ?
For me it's this : 
![image](https://github.com/user-attachments/assets/a0b902c1-e8dd-4268-af3f-5a9edb5015b7)

**Comment by damoose** [comment]
For me the command returns:

Name                        Id                                                                              Version
-------------------------------------------------------------------------------------------------------------------
Windows Subsystem for Linux MSIX\MicrosoftCorporationII.WindowsSubsystemForLinux_2.2.4.0_x64__8wekyb3d8bbwe 2.2.4.0

Except for the version it appears the same.

WSLStudio runs:

It creates the missing directories. 
It creates distributions now in //wsl$.
It creates "C:\Users\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio\lmde\installDir\ext4.vhdx"

A log file is created.

I noticed that the create CreateProjectFolders tries to log information and error messages. That seems to be a bit odd given that the logger has not been created yet.

I do see error messages in the log file though. 

2024-09-13 04:42:13.965 -07:00 [ERR] Cannot read distribution image file - Caused by exception System.IO.IOException: The process cannot access the file '\\?\C:\Users\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio\lmde2\installDir\ext4.vhdx' because it is being used by another process.
   at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.Internal.LocalFileLocator.OpenFile(String fileName, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.FileLocator.Open(String fileName, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.Vhdx.DiskImageFile..ctor(FileLocator locator, String path, FileAccess access)
   at DiscUtils.Vhdx.DiskImageFile..ctor(String path, FileAccess access)
   at WSLStudio.Helpers.WslImageHelper.ReadFile(String fileToExtract)
2024-09-13 04:42:13.965 -07:00 [ERR] Another process is already reading ext4.vhdx - Caused by exception : System.IO.IOException: The process cannot access the file '\\?\C:\Users\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio\lmde2\installDir\ext4.vhdx' because it is being used by another process.
   at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.Internal.LocalFileLocator.OpenFile(String fileName, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.FileLocator.Open(String fileName, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.Vhdx.DiskImageFile..ctor(FileLocator locator, String path, FileAccess access)
   at DiscUtils.Vhdx.DiskImageFile..ctor(String path, FileAccess access)
   at WSLStudio.Helpers.WslImageHelper.ReadFile(String fileToExtract)
   at WSLStudio.Services.DistributionInfosService.GetOsInfosFromVhdx(String distroPath, String osInfosFilePath, String osInfosPattern)
2024-09-13 04:42:13.965 -07:00 [ERR] Another process is already reading ext4.vhdx - Caused by exception : System.IO.IOException: The process cannot access the file '\\?\C:\Users\MY_USER_PROFILE_NAME\AppData\Roaming\WslStudio\lmde2\installDir\ext4.vhdx' because it is being used by another process.
   at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.Internal.LocalFileLocator.OpenFile(String fileName, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.FileLocator.Open(String fileName, FileMode mode, FileAccess access, FileShare share)
   at DiscUtils.Vhdx.DiskImageFile..ctor(FileLocator locator, String path, FileAccess access)
   at DiscUtils.Vhdx.DiskImageFile..ctor(String path, FileAccess access)
   at WSLStudio.Helpers.WslImageHelper.ReadFile(String fileToExtract)
   at WSLStudio.Services.DistributionInfosService.GetOsInfosFromVhdx(String distroPath, String osInfosFilePath, String osInfosPattern)
   at WSLStudio.Services.DistributionInfosService.GetOsInfos(String distroName, String distroPath, String field)
   
   Unclear if these are actual errors. or false positives. If false then they should not appear I think.

**Comment by NathanQuellec** [comment]
So the fix is working ? If it does well done to us :D And I will release it to the Microsoft Store as soon that you give me confirmation

- CreateProjectFolders will output log only in the console but indeed it will never be reflected in log files
- The error message is caused by the following mecanism
     - To get distribution data (like os name, os version), WSL Studio try to read the ext4.vhdx file (image of the distribution)
     - If the distribution is running (it's always the case just after the creation but it's in the background), an exception is thrown with an error in the log BUT a fallback is used (in the catch method).
     - The fallback will get data directly from the filesystem (by using //wsl$)
Maybe it's more suitable to have a WARN level for this message instead of an error.
- I don't see any improvement using pwsh.exe. In any case, since we need to install it (at least on Windows 10), it's not a good idea to use it

**Comment by NathanQuellec** [comment]
Great ! :) 
The new version of the app has been submitted to the Microsoft Store. I will post another message when it will be released

Thanks for the support, take care ! :)

**Comment by NathanQuellec** [comment]
Package released on microsoft store (v0.7.2)
